cabal-install/Distribution/Client/SetupWrapper.hs

```diff
@@ -179,6 +181,13 @@ data SetupScriptOptions = SetupScriptOptions {
     -- style with no dependencies.
     useDependenciesExclusive :: Bool,
 
+    -- | Don't add a default dependency on Cabal even if
+    -- 'useDependenciesExclusive' is @False@. Assume that it is provided via
+    -- 'useDependencies'. This is turned on (in conjunction with
+    -- useDependenciesExclusive = False) when a default setup dependencies
+    -- stanza is added by cabal-install itself. See #3199.
+    noDefaultCabalDep :: Bool,
```

Comment:
I hope we can simplify the setup wrapper in future, the logic is getting really rather hairy. Fortunately the new-build code path only needs a fraction of the functionality, since it knows exactly what deps to use in advance, so doesn't need any of the guesswork that the setupWrapper currently does.